docs/guides/haiku_migration_guide.rst

```diff
@@ -0,0 +1,192 @@
+
+Migrating from Haiku to Flax
+=============================
+
+Creating Modules
+-----------------
+
+.. codediff::
+  :title_left: Haiku
+  :title_right: Flax
+  :sync:
+
+  import haiku as hk
+
+  class Block(hk.Module):
+    def __init__(self, features: int, name=None):
+      super().__init__(name=name)
+      self.features = features
+
+    def __call__(self, x, training: bool):
+      x = hk.Linear(self.features)(x)
+      x = hk.BatchNorm(
+        create_scale=True, create_offset=True, decay_rate=0.99
+      )(x, is_training=training)
+      x = hk.dropout(hk.next_rng_key(), 0.5 if training else 0, x)
+      x = jax.nn.relu(x)
+      return x
+
+  class Model(hk.Module):
+    def __init__(self, dmid: int, dout: int, name=None):
+      super().__init__(name=name)
+      self.dmid = dmid
+      self.dout = dout
+
+    def __call__(self, x, training: bool):
+      x = Block(self.dmid)(x, training)
+      x = hk.Linear(self.dout)(x)
+      return x
+
+  ---
+
+  import flax.linen as nn
+
+  class Block(nn.Module):
+    features: int
+
+
+    @nn.compact
+    def __call__(self, x, training: bool):
+      x = nn.Dense(self.features)(x)
+      x = nn.BatchNorm(
+        use_scale=True, use_bias=True, momentum=0.99
+      )(x, use_running_average=not training)
+      x = nn.Dropout(0.5, deterministic=not training)(x)
+      x = jax.nn.relu(x)
+      return x
+
+  class Model(nn.Module):
+    dmid: int
+    dout: int
+
+
+    @nn.compact
+    def __call__(self, x, training: bool):
+      x = Block(self.dmid)(x, training)
+      x = nn.Dense(self.dout)(x)
+      return x
+
+
+.. codediff::
+  :title_left: Haiku
+  :title_right: Flax
+  :sync:
+
+  def forward(x, training: bool):
+    return Model(256, 10)(x, training)
+
+  model = hk.transform_with_state(forward)
+
+  ---
+
+  ...
+
+
+  model = Model(256, 10)
+
+Initialization
+---------------
+
+.. codediff::
+  :title_left: Haiku
+  :title_right: Flax
+  :sync:
+
+  sample_x = jax.numpy.ones((1, 784))
+  params, state = model.init(
+      PRNGKey(0),
+      sample_x, training=False # <== inputs
```

Comment:
Why is `training=False` here, but `training=True` for Flax?

Reply:
In Flax we commonly initialize with `training=False` so Dropout behaves deterministically and doesn't request a PRNGKey. However, Haiku's BatchNorm layer doesn't seem to like to be initialize with `training=False` so I had to turn it to `True`. I'll be sure to mention this in the text.